### ISSUE TITLE
After post-training the control model, a noise issue has occurred.

### ISSUE THREAD

**Issue by YirunLi** [nvidia-cosmos/cosmos-transfer1#87]
After training Cosmos-Transfer1-7B, I merged the models and the inference results were as follows.
![Image](https://github.com/user-attachments/assets/b9faeebb-8da6-4425-8a97-25ac7a1b006a)
Firstly, according to https:Following the instructions of github.com/nvidia-cosmos/cosmos-transfer1/blob/main/examples/training_cosmos_transfer_7b.md, the data was preprocessed, T5 Text Embeddings were generated, and base_model.pt and depth_control.pt were split using the scripts/convert_ckpt_fsdp_to_tp.py script.

Then, 175,000 iter were trained on 8 x H100s. After training, the tp model was merged into iter_000175000_reg_model.pt and iter_000175000_ema_model_only.pt using the convert_ckpt_tp_to_fsdp.py script.

According to the description in the paper, base_model is frozen during training, so I only replaced the model of Cosmos-Transfer1-7B/depth_control.pt with iter_000175000_ema_model_only.pt that I just merged, and this result appeared.

I ran the inference process in a conda environment on a single GPU.
And I tried training with 10 iters as suggested in https:github.com/nvidia-cosmos/cosmos-transfer1/issues/77. I found that the result is the same as it inferenced by the origin model, indicating that 10iters was too few iterations and did not change the original model.

So how should I locate my problem?

**Comment by Rxliang** [comment]
Met the same issue with Depth control

**Comment by qianlim** [comment]
Thanks for spotting this. We are looking into this and will get back to you.

**Comment by Rxliang** [comment]
Hi @qianlim  Thanks for looking into this. I think the edge_control training is fine but depth and seg both failed

**Comment by qianlim** [comment]
@Rxliang Thanks for this information! Just to confirm: "edge_control training is fine" -- do you mean training edge_control for a similar number of steps, merge model, run inference with the merged model, and the results make sense?

If that's the case, we can rule out the problem with the checkpoint conversion code, architecture and the training infra code and I'll focus investigating the depth/seg control part.

**Comment by Rxliang** [comment]
@qianlim I tried to train it longer for edge_control to 5k iters it also began to corrupt the inference results started to show the sign of noise-like generation and the temporal consistancy showed some level of degradation. The loss seems to drop to the level of 2.5, is there a loss curve for reference? I did not train it such long iters like @YirunLi did. 

Wheras for depth and seg, I took 100 iters trained and merged 15.14gb reg model pt for inference use. like discussed here https://github.com/nvidia-cosmos/cosmos-transfer1/issues/49. These both gave noisy inference generation however the same merging process yields better results for edge_control 5k iters ckpts (yet still not good enough when compared with pre-trained weights.)

**Comment by qianlim** [comment]
Found one issue. We tried the following:
Using the pretrained (good) edge checkpoint, convert to TP, and then converted back to FSDP checkpoint. Then run inference with this converted checkpoint. Got a similar noisy output as above. This indicates at least there is some issue with the conversion code. Looking further into it now.

**Comment by Rxliang** [comment]
Hi is there any update on this thread? Looking forward to hearing from you

**Comment by Rxliang** [comment]
@pjannaty Unfortunately no. Still noisy on my end I tried the reg model whose size was 1.66Gb which looks normal when compared with my previous issue @qianlim looked into Here (https://github.com/nvidia-cosmos/cosmos-transfer1/issues/49) Have you testd and tried internally first? I can also try to merge the pre-trained weights directly to see if it's better.

https://github.com/user-attachments/assets/4c783e3d-70b4-4182-95fa-9f95db584017

**Comment by leungjch** [comment]
I'm also able to reproduce the same noisy output from @Rxliang even after applying the fix from #111 . Steps to reproduce (mostly just following the post-training README)
- Prepare a dataset of videos, text embeddings, and captions (I'm using edge_control so no need for custom inputs).
- Split the base_model.pt and edge_control.pt checkpoints into TP checkpoints ([link](https://github.com/nvidia-cosmos/cosmos-transfer1/blob/main/examples/training_cosmos_transfer_7B_sample_AV.md#2-splitting-the-checkpoints-to-tensorparallel-checkpoints))
- Run post-training: e.g. `torchrun --nproc_per_node=8 -m cosmos_transfer1.diffusion.training.train --config=cosmos_transfer1/diffusion/config/config_train.py -- experiment=CTRL_7Bv1pt3_lvg_tp_121frames_control_input_edge_block3_posttrain`
- Copy the checkpoints from `cosmos_transfer1_posttrain/CTRL_7Bv1_lvg/CTRL_7Bv1pt3_lvg_tp_121frames_control_input_edge_block3_posttrain/fsdp_checkpoints/edge_control_reg_model.pt` (or  into `cosmos-transfer1/checkpoints/nvidia/Cosmos-Transfer1-7B/edge_control.pt` and run inference using edge_control

I was able to verify the fix in #111 fixes the conversion (i.e. convert original edge_control.pt into TP -> convert to FSDP -> run inference and verify the output is good). So it seems that the root cause of the post-training noise issue still remains to be solved.

@pjannaty Would be happy to provide the dataset to the Cosmos team to help repro if needed, thanks 🙏

**Comment by dedoogong** [comment]
I met the same issue with my custom canny edge models!

**Comment by Jay-Ye** [comment]
I got similar results. I can confirm that the model conversion script between TP and FSDP is working correctly after applying the fix of #111.

Here's what I tried:
I first converted the pretrained vis checkpoint to TP, ran training for 200 iterations, then converted the saved TP models to FSDP checkpoint and ran inference. The results were similar to the original pretrained model. However, as training continued, the results got worse. After about 1k iterations, I ended up with a noisy video like above.

The issue might be related to the training configs. I also visualized the predicted video from the first training iteration (following the code from [cosmos-predict1/issues/97](https://github.com/nvidia-cosmos/cosmos-predict1/issues/97)). I also checked the `extra_per_block_abs_pos_emb` value as mentioned in that issue, and it remains the same during training and inference. Here is what I got:

https://github.com/user-attachments/assets/3a14955f-70f0-4c70-bbb8-5863e39084ce

**Comment by guopengf** [comment]
I'm facing the same issue.

**Comment by qianlim** [comment]
I'm further narrowing down the issue.

We have tried training a Transfer1-AV-Sample (single view) from this public repo on ~1000 clips, on 8 nodes. Using the same default hyperparameter configs as in this repo. It has the same architecture as Edge Control. It yields reasonable results.
This means the training infra code and the checkpoint conversion code works fine.

Now i'm suspecting the batch size can be an issue. In the current code, batch size is set to 1 due to gpu memory limits. It can be problematic. In our internal testings we always used several nodes (which means effective batch size = number of nodes). I'm now running a 1-node job to further debug.

In the mean time, @guopengf @Jay-Ye @dedoogong @leungjch @Rxliang What's the batch size you guys use for training the model? Did you change the `grad_accum_iter` (gradient accumulation, which affects the effective batch size) parameter in the trainer config? And how many nodes (1 node = 8 gpus) do you use for training?

**Comment by dedoogong** [comment]
hi @qianlim , in case of me, I'm using 1 node(8 H100 GPUs) for training (if I see the pre/post training really work, I will extend to use 2 nodes, 16 GPUs)

**Comment by Jay-Ye** [comment]
Hi @qianlim, thanks so much for looking into this issue.
In my test, I exactly followed the [post-training instructions](https://github.com/nvidia-cosmos/cosmos-transfer1/blob/main/examples/training_cosmos_transfer_7b.md) in this repo without any modifications to fine-tune cosmos-transfer1[vis] on a single node.

How many iterations did you train? In my case, the results from a model post-trained for 200 iterations looked reasonable, but were getting worse till pure noise as training continued.

Also see my comment above:
> Here's what I tried: I first converted the pretrained vis checkpoint to TP, ran training for 200 iterations, then converted the saved TP models to FSDP checkpoint and ran inference. The results were similar to the original pretrained model. However, as training continued, the results got worse. After about 1k iterations, I ended up with a noisy video like above.

Could you please try to visualize the predicted videos during the training loop and see if the results are reasonable?
You can simply add the following code snippet after this line.
https://github.com/nvidia-cosmos/cosmos-transfer1/blob/e4055e39ee9c53165e85275bdab84ed20909714a/cosmos_transfer1/diffusion/training/models/model_image.py#L229
```
        # NOTE just for debugging
        if torch.distributed.get_rank() == 0:
            original_video = self.decode(x0) # self.vae.decode(x0)
            predicted_video = self.decode(model_pred.x0)
            import imageio
            import numpy as np
            # Create video writer
            writer = imageio.get_writer('debug_vis.mp4', fps=24)
            # Write frames side by side
            for i in range(original_video.shape[2]):
                orig_frame = original_video[0, :, i, :, :]
                pred_frame = predicted_video[0, :, i, :, :]
                orig_frame = ((1.0 + orig_frame).clamp(0, 2) * 127.5).to(torch.uint8).permute(1, 2, 0).cpu().numpy()
                pred_frame = ((1.0 + pred_frame).clamp(0, 2) * 127.5).to(torch.uint8).permute(1, 2, 0).cpu().numpy()
                combined_frame = np.concatenate([orig_frame, pred_frame], axis=1)
                writer.append_data(combined_frame)
            writer.close()
        exit()
```

Ideally, the predicted video should be realistic instead of noisy since we're post-training from a well-trained model. However, in my case, I can only see pure noise even at the beginning of the training process.

**Comment by guopengf** [comment]
Hi @Jay-Ye and @dedoogong, meanwhile, do you also observe relatively large loss value? In my experiment, the loss value is in the range of [4.0, 6.0]. It is quite large compared to my fine-tuning experience using the Cosmos-predict model.

**Comment by qianlim** [comment]
Hi @Jay-Ye @dedoogong @Rxliang @guopengf @leungjch -- I found a solution that fixes the issue on my side. Could you try out and see if it works for you? Steps see below. Would appreciate your help with the verification.

The key is this line: https://github.com/nvidia-cosmos/cosmos-transfer1/pull/149/files#r2176350739

Steps: 
- Option 1: You can directly checkout the commit `626cd1` in [PR149](https://github.com/nvidia-cosmos/cosmos-transfer1/pull/149/commits).
- Option 2: Alternatively, you can add that line's change (` base_state_dict = {k.replace('base_model.', ''): v for k, v in base_state_dict.items()}`) to your current code's `load_base_model` function.
- No need to change batch size, number of nodes etc. 
- Launch a simple training. I tested with 200 training examples. 
- If the first iteration loss is 4.x or even larger, the training will very likely fail. If the first few iterations are with loss of 0.x or even negative values, it should be good. 

Looking forward to your feedbacks. Thanks!

**Comment by guopengf** [comment]
I can confirm that this fix works on my end.

**Comment by leungjch** [comment]
I tried training edge_control for 1000 iters and don't seem to get a noisy generation result using it anymore as well, so I believe the fix works. Thank you!
